pkg/blobstore/spanner_blob_access.go

```diff
@@ -0,0 +1,975 @@
+			s,
+		},
+	})
+	if err != nil {
+		return err
+	}
+	if err = op.Wait(ctx); err != nil {
+		return err
+	}
+	return nil
+}
+
+func createGCSBucket(ctx context.Context, gcsBucket *storage.BucketHandle, databaseName string, daysToLive uint64) error {
+	// Set up the deletion policy.
+	var attrs *storage.BucketAttrs
+
+	// If daysToLive is zero, use the default.
+	if daysToLive == 0 {
+		daysToLive = defaultDaysToLive
+	}
+
+	lifecycle := storage.Lifecycle{}
+	lifecycle.Rules = append(lifecycle.Rules, storage.LifecycleRule{
+		Action: storage.LifecycleAction{
+			Type: storage.DeleteAction,
+		},
+		Condition: storage.LifecycleCondition{
+			DaysSinceCustomTime: int64(daysToLive),
+		},
+	})
+	attrs = &storage.BucketAttrs{
+		Lifecycle: lifecycle,
+	}
+
+	// Extract the project ID from the spanner database name (no, really).
+	s := strings.Split(databaseName, "/")
+	projectID := s[1]
+
+	// Create the bucket.
+	if err := gcsBucket.Create(ctx, projectID, attrs); err != nil {
+		return err
+	}
+	return nil
+}
+
+func getGCSTTL(ctx context.Context, gcsBucket *storage.BucketHandle) (uint64, error) {
+	attrs, err := gcsBucket.Attrs(ctx)
+	if err != nil {
+		return 0, err
+	}
+	for _, r := range attrs.Lifecycle.Rules {
+		if r.Action.Type == storage.DeleteAction && r.Condition.DaysSinceCustomTime != 0 {
+			return uint64(r.Condition.DaysSinceCustomTime), nil
+		}
+	}
+	return 0, nil
+}
+
+func updateGCSDeletionPolicy(ctx context.Context, gcsBucket *storage.BucketHandle, daysToLive uint64) error {
+	lifecycle := storage.Lifecycle{}
+	lifecycle.Rules = append(lifecycle.Rules, storage.LifecycleRule{
+		Action: storage.LifecycleAction{
+			Type: storage.DeleteAction,
+		},
+		Condition: storage.LifecycleCondition{
+			DaysSinceCustomTime: int64(daysToLive),
+		},
+	})
+	attrs := storage.BucketAttrsToUpdate{
+		Lifecycle: &lifecycle,
+	}
+	_, err := gcsBucket.Update(ctx, attrs)
+	return err
+}
+
+// Convert a digest to the key of the entry in the Spanner database and GCS bucket.
+func (ba *spannerBlobAccess) digestToKey(digest digest.Digest) string {
+	sz := strconv.FormatInt(digest.GetSizeBytes(), 10)
+	if ba.storageType == pb.StorageType_ACTION_CACHE {
+		// Instance names are hierarchical, but '/' has special meaning for GCS.
+		// We don't need a hierarchical namespace for storage.
+		instance := strings.ReplaceAll(digest.GetInstanceName().String(), "/", "-")
+		return digest.GetHashString() + "-" + sz + "-" + instance
+	} else if ba.storageType == pb.StorageType_CASTORE {
+		return digest.GetHashString() + "-" + sz
+	} else {
+		// Code in pkg/blobaccess/configuration/new_blob_access.go should prevent this, but
+		// since that lives in a separate place, let's be sure.
+		panic("Invalid Spanner storage Type configured")
+	}
+}
+
+func (ba *spannerBlobAccess) keyToBlobSize(key string) (int64, error) {
+	s := strings.Split(key, "-")
+	sz, err := strconv.ParseInt(s[1], 10, 64)
+	if err != nil {
+		return -1, err
+	}
+	return sz, nil
+}
```

Comment:
How does this hold for the Action Cache? In the case of the Action Cache the Action Digest pertains to the size of the Action message. Not the ActionResult.

Reply:
Oh, nice catch.  It's basically dumb luck that this doesn't cause problems, because action cache entries and action messages are small, so we end up using Spanner.  The only place I validate the buffer size is in the CAS.

Reply:
I rewrote the code to use the knowledge of where the blobs live.  On the write side, we use the size to determine where the blob should be stored.  On the read side, we use the existence of InlineData in the Spanner record to determine where the blob lives.  I have an MR out to (optionally) write small blobs to GCS also so that they can be accessed from GCS-FUSE.  As soon as I get verification that it works, I can send a PR upstream.